### ISSUE TITLE
Out of memory when downloading file.

### ISSUE THREAD

**Issue by Yayo-Arellano** [kittinunf/Fuel#108]
When i download a big file I am getting an out of memory error in Android. If the file is small there is no problem.

Maybe it can be related to this. 

[Out of memory error](http://stackoverflow.com/questions/8730269/out-of-memory-when-downloading-a-file) or this [Out of memory error 2](http://stackoverflow.com/questions/4045028/android-download-file-out-of-memory-woes)

**Comment by Yayo-Arellano** [comment]
I test using the solution [solution](http://stackoverflow.com/a/12263724/3479489) from stakoverflow and the out of memory error is gone. Hope it helps.

**Comment by kittinunf** [comment]
Yeah, you are right. Would you mind open a PR to fix that? Thanks!

**Comment by Yayo-Arellano** [comment]
Sorry. what i did was testing the method i just commented. I did not change anything in the library.  I am not really sure how to fix it in the library.

**Comment by Yayo-Arellano** [comment]
I tried to update the library. But for some reason i can not test it.  
`class DownloadTaskRequest(request: Request) : TaskRequest(request) {

    val BUFFER_SIZE = 1024

    var progressCallback: ((Long, Long) -> Unit)? = null
    lateinit var destinationCallback: ((Response, URL) -> File)

    lateinit var inputStream: InputStream
    lateinit var fileOutputStream: FileOutputStream
    lateinit var bufferedInputStream: BufferedInputStream

    override fun call(): Response {
        val response = super.call()
        val file = destinationCallback.invoke(response, request.url)
        //file output
        fileOutputStream = FileOutputStream(file)
        inputStream = ByteArrayInputStream(response.data)
        bufferedInputStream = BufferedInputStream(inputStream, BUFFER_SIZE);

        var byteArr = ByteArray(BUFFER_SIZE);
        var progress: Long = 0;
        var actual: Int = 0;
        while (actual != -1) {
            fileOutputStream.write(byteArr, 0, actual);
            actual = bufferedInputStream.read(byteArr, 0, BUFFER_SIZE);
            progress += actual;
            progressCallback?.invoke(progress, response.httpContentLength);
        }
        fileOutputStream.close();
        bufferedInputStream.close();
        inputStream.close();
        
        return response
    }

}`

**Comment by kittinunf** [comment]
Have your tried running test with `./gradlew test --info` on command line?

**Comment by dstd** [comment]
@Yayo-Arellano is it a final version of your update of `DownloadTaskRequest`? I can't understand what you think you have fixed.

As I understand the code of Fuel, it always reading the whole response body into the memory.
So if your app has not so much free memory left and makes a request for a large file (eg. 30MB) you can run out of memory, because all these 30MBs will be stored in memory as a `byte[]`.

The solution specified in your 2nd message has a different approach – it reads `connection.inputStream` by a small chunks of bytes, so it is not OOM-prone.

Maybe @kittinunf can give a suggestion how to fix the library so the data will be processed as a stream without storing in the memory.

**Comment by yoavst** [comment]
```kotlin
fun fileDeserializer(file: File) = object : ResponseDeserializable<Unit> {
    override fun deserialize(inputStream: InputStream) {
        inputStream.copyTo(file.outputStream())
    }
}
```

**Comment by dstd** [comment]
@yoavst you're missing one important thing – all these deserializable variants are uses a data stored in RAM. In fact `inputStream` will be a `ByteArrayInputStream`, so that a large file can cause OOM.

Maybe it's me who can't see the code that handles `connection.inputStream` as a stream, without reading all the content into the memory. Correct me in this case, please

**Comment by Yayo-Arellano** [comment]
@dstd  I did not update the library. I used the solution in this [answer](http://stackoverflow.com/questions/4045028/android-download-file-out-of-memory-woes) and now i can download 500MB files with out a crash. and before even in 50MB files the app will have the OOM. The code i post was some attempt to fix the library but i did not succeed.

**Comment by kittinunf** [comment]
I am sorry for late to the party.

I just wanna get to the bottom of the problem.

Well, let's see, this is the gist of function that Fuel uses to download the file.

https://github.com/kittinunf/Fuel/blob/a702a39043519d2b3e06c69a3afd10bec2b24ffe/fuel/src/main/kotlin/com/github/kittinunf/fuel/core/requests/DownloadTaskRequest.kt#L28

```
fun InputStream.copyTo(out: OutputStream, bufferSize: Int = DEFAULT_BUFFER_SIZE, progress: ((Long) -> Unit)?): Long {
    var bytesCopied = 0L
    val buffer = ByteArray(bufferSize)
    var bytes = read(buffer)
    while (bytes >= 0) {
        out.write(buffer, 0, bytes)
        bytesCopied += bytes
        progress?.invoke(bytesCopied)
        bytes = read(buffer)
    }
    return bytesCopied
}
```
If I understand correctly, there is one real key difference. Fuel implementation uses default_buffer_size which is 8 MB (it was a constant val from stdlib) but above implementation uses only 1 MB.

So if we just passed smaller bufferSize into a function and it should be also fixed the problem? @Yayo-Arellano Could you please help me try that out?

**Comment by dstd** [comment]
@kittinunf take a look few lines above `copyTo`:
```
    override fun call(): Response {
        val response = super.call()
        val file = destinationCallback.invoke(response, request.url)
        //file output
        fileOutputStream = FileOutputStream(file)
        dataStream = ByteArrayInputStream(response.data)
        dataStream.copyTo(fileOutputStream, BUFFER_SIZE) { readBytes ->
            progressCallback?.invoke(readBytes, response.httpContentLength)
        }
        return response
    }
```
Here `response.data` contains all the data received from the server, all 500 MBs.
Just because of this block of code in [HttpClient.kt#L55](https://github.com/kittinunf/Fuel/blob/a702a39043519d2b3e06c69a3afd10bec2b24ffe/fuel/src/main/kotlin/com/github/kittinunf/fuel/toolbox/HttpClient.kt#L55):
```
    if (dataStream != null) {
        data = if (contentEncoding.compareTo("gzip", true) == 0) {
            GZIPInputStream(dataStream).readBytes()
        } else {
            dataStream.readBytes()
        }
    }
```

**Comment by kittinunf** [comment]
@dstd Yeah, you are right. What I am confused is the fact that @Yayo-Arellano's implementation is irrelevant with code from `HttpClient`. So I failed to see why above implementation makes him downloading 500 MB of data just fine.

**Comment by Yayo-Arellano** [comment]
Sorry. Maybe i did not explain well what i was trying to say. 
I did use the code from `HttpClient` but not using this library, not using kotlin. So it fixes "My" problem but not the library bug. Then i try to update the library with the solution i use translating the solution in this[ link ](http://stackoverflow.com/a/12263724/3479489) to kotlin.

**Comment by dstd** [comment]
Here is a patch that gives the support of I/O streams to handle the upload/download. 
As a side effect it makes a progress reporting to work correctly =)

In my project I stuck on older version of Fuel due to several app-specific changes and still can't find large enough pile of time to jump into `master`. So probably this code can't be used as is.

Main magic is in `HttpClient.kt`, other changes is mainly about setting it up, and few changes about form-data handling that has been fixed in `master` as well, but in a bit different way.

https://gist.github.com/dstd/cc5710c52c58eec6f14ee1eb229eb4de

**Comment by JackED42** [comment]
Hi, 

Are there any plans on adding this fix to an upcoming version?

Thanks

Jack

**Comment by kittinunf** [comment]
Yup, I will made a lot of fixes in next version. I haven't looked for this yet, but I will soon. In the meantime, if you wants to help me out, it is appreciated, too!

**Comment by westlinkin** [comment]
Any update?

**Comment by kittinunf** [comment]
I think this should be fixed with the latest version of Fuel 🎉

**Comment by charshep** [comment]
Correct me if I'm wrong. I'm new to Kotlin and even Java I haven't touched in probably 10 years. But don't lines `DownloadTaskRequest.kt:22` and `InputStreams.kt:19` ensure that the entire file will be read into memory due to the call to `byteBuffer.toByteArray()`?


    override fun call(): Response {
        val response = super.call()
        val file = destinationCallback(response, request.url)

        FileOutputStream(file).use {
            response.dataStream.copyTo(out = it, bufferSize =  BUFFER_SIZE, progress =  { readBytes ->
                progressCallback?.invoke(readBytes, response.contentLength)
            }){
 22               **response.data = it**
            }
        }
        return response
    }
==========

**Comment by charshep** [comment]
Correct me if I'm wrong, I'm new to Kotlin and even Java I've been away from for quite a while. But don't `DownloadTaskRequest.kt:22` and `InputStreams.kt:19` result in the entire downloaded file contents being read into memory?

    override fun call(): Response {
        val response = super.call()
        val file = destinationCallback(response, request.url)

        FileOutputStream(file).use {
            response.dataStream.copyTo(out = it, bufferSize =  BUFFER_SIZE, progress =  { readBytes ->
                progressCallback?.invoke(readBytes, response.contentLength)
            }){
    22        **response.data = it**
            }
        }
        return response
    }

and

    internal fun InputStream.copyTo(out: OutputStream, bufferSize: Int = DEFAULT_BUFFER_SIZE, progress: ((Long) -> Unit)?, onComplete:((ByteArray)->Unit)? = null): Long {
    var bytesCopied = 0L
    val buffer = ByteArray(bufferSize)
    var bytes = read(buffer)
    val byteBuffer = ByteArrayOutputStream(bufferSize)
    while (bytes >= 0) {
        out.write(buffer, 0, bytes)
    14    **byteBuffer.write(buffer,0,bytes)**
        bytesCopied += bytes
        progress?.invoke(bytesCopied)
        bytes = read(buffer)
    }
    19    **onComplete?.invoke(byteBuffer.toByteArray())**
    return bytesCopied
    }

**Comment by markGilchrist** [comment]
@charshep This was a change that I think I added and the reason I did so was we had a situation where when downloading data to a file it was not then accessible in the Result<byteArray,Error>.

I did test this with then mocking lib (httpBin) that we were using at the time and I remember that large byteArrays, when requested from httpBin.org were caped by the server. so I think I didn't notice this at the time.

I think that you right both in reporting and diagnosing this problem. I have just merged in my latest pr so I am happy to look and implement another solution unless you would like to pick this up and raise a pr?
